test/integration/client-side-encryption/client_side_encryption.prose.test.js

```diff
@@ -44,6 +55,412 @@ describe('Client Side Encryption Prose Tests', function () {
+          expect(e.message).to.include('expected UTF-8 key');
+        }
+      });
+    });
+
+    // Case 2.
+    context('when using azure', function () {
+      const masterKey = {
+        keyVaultEndpoint: 'doesnotexist.local',
+        keyName: 'foo'
+      };
+
+      it('fails with no tls', async function () {
+        try {
+          await clientEncryptionNoTls.createDataKey('azure', { masterKey });
+        } catch (e) {
+          expect(e.originalError.message).to.include('certificate required');
+        }
+      });
+
+      it('fails with invalid host', async function () {
+        try {
+          await clientEncryptionWithTls.createDataKey('azure', { masterKey });
+        } catch (e) {
+          expect(e.message).to.include('HTTP status=404');
+        }
+      });
+
+      it('fails with expired certificates', async function () {
+        try {
+          await clientEncryptionWithTlsExpired.createDataKey('azure', { masterKey });
+        } catch (e) {
+          expect(e.originalError.message).to.include('certificate has expired');
+        }
+      });
+
+      it('fails with invalid hostnames', async function () {
+        try {
+          await clientEncryptionWithInvalidHostname.createDataKey('azure', { masterKey });
+        } catch (e) {
+          expect(e.originalError.message).to.include('does not match certificate');
+        }
+      });
+    });
+
+    // Case 3.
+    context('when using gcp', function () {
+      const masterKey = {
+        projectId: 'foo',
+        location: 'bar',
+        keyRing: 'baz',
+        keyName: 'foo'
+      };
+
+      it('fails with no tls', async function () {
+        try {
+          await clientEncryptionNoTls.createDataKey('gcp', { masterKey });
+        } catch (e) {
+          expect(e.originalError.message).to.include('certificate required');
+        }
+      });
+
+      it('fails with invalid host', async function () {
+        try {
+          await clientEncryptionWithTls.createDataKey('gcp', { masterKey });
+        } catch (e) {
+          expect(e.message).to.include('HTTP status=404');
+        }
+      });
+
+      it('fails with expired certificates', async function () {
+        try {
+          await clientEncryptionWithTlsExpired.createDataKey('gcp', { masterKey });
+        } catch (e) {
+          expect(e.originalError.message).to.include('certificate has expired');
+        }
+      });
+
+      it('fails with invalid hostnames', async function () {
+        try {
+          await clientEncryptionWithInvalidHostname.createDataKey('gcp', { masterKey });
+        } catch (e) {
+          expect(e.originalError.message).to.include('does not match certificate');
+        }
+      });
+    });
+
+    // Case 4.
+    context('when using kmip', function () {
+      it('passes with tls', async function () {
+        const dataKey = await clientEncryptionWithTls.createDataKey('kmip');
+        // TODO: NODE-3927
+        await clientWithTls.close();
```

Comment:
This test leaks a session when the client close is in an `after` instead of an `afterEach`, but since the other tests are failure scenarios the client is needed to be open. When putting in an `afterEach` we noticed in this case the `afterEach` was executing before the `it` finished, closing the client before is was getting used. Opened NODE-3927 to investigate later.

Reply:
Just removed as the same thing is getting tested in the kmip context above.